### PR TITLE
Trigger a deprecation when setting a deprecated contao.localconfig key

### PR DESCRIPTION
**Pull request by m-vo**
This makes sure setting deprecated `localconfig` keys via the bundle configuration will actually trigger a deprecation.

see https://github.com/contao/contao/issues/3965#issuecomment-1021503954

### CODE REVIEW

File: core-bundle/src/Resources/contao/library/Contao/Config.php [hunk 1]
```diff
@@ -408,10 +408,7 @@ public static function has($strKey)
 	 */
 	public static function get($strKey)
 	{
-		if (isset(self::$arrDeprecated[$strKey]) || isset(self::$arrDeprecatedMap[$strKey]))
-		{
-			trigger_deprecation('contao/core-bundle', '4.12', 'Using "%s(\'%s\')" has been deprecated. Use the "%s" parameter instead.', __METHOD__, $strKey, self::$arrDeprecated[$strKey] ?? self::$arrDeprecatedMap[$strKey]);
-		}
+		self::checkForDeprecation($strKey, sprintf('Using "%s(\'%s\', …)"', __METHOD__, $strKey));
```
Comment (ausi) [inline core-bundle/src/Resources/contao/library/Contao/Config.php]:
```suggestion
		self::checkForDeprecation($strKey, sprintf('Using "%s(\'%s\')"', __METHOD__, $strKey));
```

**Review by ausi — DISMISSED** [review]

**Review by Toflar — DISMISSED** [review]

**Review by bytehead — DISMISSED** [review]

**Review by ausi — APPROVED** [review]

**Review by aschempp — APPROVED** [review]

**Review by leofeyer — CHANGES_REQUESTED** [review]
I think this is over-engineered. IMHO we should

* keep the deprecation check in the `Config::get()` method;
* not trigger a deprecation in the `Config::set()` method;
* maybe trigger a deprecation in the config tree builder but without adding another public static method to the `Config` class.

**Comment by ausi** [comment]
> not trigger a deprecation in the `Config::set()` method;

But we already do that in the current version. I don’t think we should remove that, should we?
Related: #3056

**Comment by leofeyer** [comment]
My bad. Of course we should keep it then.

If we already trigger a deprecation when setting or getting the key, why do we need another deprecation in the config tree at all?

**Comment by ausi** [comment]
> why do we need another deprecation in the config tree at all?

I think because it would get triggered at container build time and should therefore be seen much more likley by a developer.

**Comment by leofeyer** [comment]
I see. However, I do not like it very much. The configuration tree should not have to know about expired config keys. Maybe we should use `Config::set()` here instead?

https://github.com/contao/contao/blob/2ba558f74e0183ed2daec842f01198dd89e89b1b/core-bundle/src/DependencyInjection/ContaoCoreExtension.php#L82

**Comment by ausi** [comment]
> Maybe we should use `Config::set()` here instead?

We could call `Config::get()` there, just to trigger the deprecation warning (without changing $GLOBALS).

**Comment by m-vo** [comment]
> We could call Config::get() there, just to trigger the deprecation warning (without changing $GLOBALS).

I'm ok with that if you prefer, although I think not using the results of a non-void function and assuming some behavior is way less expressive than just using a function made for this scenario.

**Comment by leofeyer** [comment]
Yes, if we really have to show the deprecations during the container build, this would be the least invasive change. I still don‘t see any added value though.

**Comment by m-vo** [comment]
> I still don‘t see any added value though.

https://github.com/contao/contao/issues/3965#issuecomment-1021503954

You won't get any deprecation notices when setting deprecated config keys via the `contao.localconfig`, but IMHO you should. That we did not trigger deprecations like with the other config keys, is due to the fact we're using a variable node here. But that's only an implementation detail.

**Comment by m-vo** [comment]
Just checked if we could use `Config::get()` but now I remember why I didn't use that. 

This will produce a message like so:

```
Since contao/core-bundle 4.12: Using "Contao\Config::get('enableSearch')" has been deprecated. Use the "contao.search.default_indexer.enable" parameter instead.
```

But it should rather contain the respective config key:
```
Since contao/core-bundle 4.12: Setting "contao.localconfig.enableSearch" has been deprecated. Use the "contao.search.default_indexer.enable" parameter instead.
```

**Comment by leofeyer** [comment]
I have created another PR with a different approach: #4076

**Comment by leofeyer** [comment]
Closed in favor of #4076.

**Pull request closed without merging**